jdbi/src/main/java/org/killbill/commons/jdbi/guice/DataSourceProvider.java

```diff
@@ -230,6 +243,57 @@ DataSource buildDataSource() {
         }
     }
 
+    private class UnpooledDataSourceBuilder {
+
+        DataSource buildDataSource() {
+            String dataSourceClassName = config.getDataSourceClassName();
+            if (Strings.isNullOrEmpty(dataSourceClassName)) {
+                throw new IllegalArgumentException("dataSourceClassName was not specified in the configuration and could not be inferred");
+            }
+
+            if (DatabaseType.MYSQL.equals(databaseType)) {
+                Reference ref = new Reference(dataSourceClassName);
+                ref.add(new StringRefAddr("user", config.getUsername()));
+                ref.add(new StringRefAddr("password", config.getPassword()));
+                ref.add(new StringRefAddr("explicitUrl", "true"));
+                ref.add(new StringRefAddr("url", config.getJdbcUrl()));
+                try {
+                    return (DataSource) new MysqlDataSourceFactory().getObjectInstance(ref, null, null, null);
```

Comment:
> `HikariPool` is a `final` class whose methods I can't override to behave the way I want them to.

Ah indeed, I overlooked this 😿 

> Or are you saying I should use the Javassist library to [specify my own implementations for the `ProxyFactory` methods](https://github.com/brettwooldridge/HikariCP/blob/HikariCP-2.3.7/hikaricp/src/main/java/com/zaxxer/hikari/proxy/JavassistProxyFactory.java#L95-L120) at runtime?  I'm not clear on how that works, but I could look into it if you think that's a promising approach.

Nah, that's probably too much of a headache.

> Re MySQL's `serverName` property:  since I provide the `url` property, am I not allowed to omit `serverName`?  I admit I haven't tested that yet.

My remark was more about the risk of missing or misconfiguring the `DataSource`, rather than this specific property. We have several pools and layers of `DataSource` in the platform code, and I'm always nervous whenever we need to touch this, especially when we need to instantiate/manipulate drivers code directly. Hence my pushback to see if there is a way we could simply this.

Let me suggest one more route.

Re-reading the code, `DataSourceProvider` is injected via `ReferenceableDataSourceSpyProvider` in [KillbillPlatformModule](https://github.com/killbill/killbill-platform/blob/master/server/src/main/java/org/killbill/billing/server/modules/KillbillPlatformModule.java#L115) and in [DefaultOSGIModule](https://github.com/killbill/killbill-platform/blob/master/osgi/src/main/java/org/killbill/billing/osgi/glue/DefaultOSGIModule.java#L86). We already have access to [EmbeddedDB](https://github.com/killbill/killbill-commons/blob/master/embeddeddb/common/src/main/java/org/killbill/commons/embeddeddb/EmbeddedDB.java) (confusing name - it's really just an abstraction for the DB) in the first module (where it's bound) and it's easy to pass it to `DefaultOSGIModule` (because this module is initiated from `KillbillPlatformModule`).

This means that `EmbeddedDB` could be injected to `DataSourceProvider` and could provide the `DataSource` for free, including both MySQL and MariaDB (`initialize` might need to be called manually though, TBD). In case of PostgreSQL, the code would return `PGSimpleDataSource` for instance. Would that work?